### PR TITLE
Add logging for learning rates in MetricsProcessor

### PR DESCRIPTION
**Pull request by idoh**
This PR adds learning rate logging. There was a previous attempt to implement this in an [earlier PR](https://github.com/pytorch/torchtitan/pull/937), but that one was ultimately **closed**. This version ensures that LR logging works properly, I verified it using the WSD scheduler that was recently added in [another PR](https://github.com/pytorch/torchtitan/pull/938).
image

One design consideration here is that torchtitan supports multiple optimizers and learning rate schedules, each potentially having its own LR. However, in practice, I believe that 99.9999% of use cases will use a single LR. 

Given that, the logging works as follows:
- If there is only one learning rate, it gets logged directly under the main charts as `lr`.
- If there are multiple learning rates, they are logged under a separate section, each with its corresponding label.

Alternatively, we could have ignored the multi-LR case and always logged a single LR, but I prefer this approach since it handles both scenarios robustly with minimal extra code.

Happy to adjust if others have a strong preference for simplicity over robustness.

### CODE REVIEW

**Comment by idoh** [comment]
@wwwjn I think this can be super useful for researchers as the convergence curves and learning rate are highly correlated and this helps visualize the learning rate. Are there any additional checks you would like me to run?

**Comment by wwwjn** [comment]
Thanks for making this PR! This PR look good to me, will let @tianyu-l  to take another look before merging

**Comment by idoh** [comment]
Hey @tianyu-l, I noticed that there is a [new PR](https://github.com/pytorch/torchtitan/pull/1453) on this same issue, so I'm guessing that many users need this feature. Your feedback and help in pushing this PR would be greatly appreciated.

**Review by tianyu-l — CHANGES_REQUESTED** [review]
sorry for the delay in reviewing

Let's just log `self.lr_schedulers.schedulers[0].get_last_lr()[0]`.
Currently in torchtitan
- `schedulers[0]` gives the lr scheduler for the model part from first PP stage (if there are multiple model parts).
- `get_last_lr()[0]` gives the first lr corresponding to the first optimizer group within an optimizer, which currently each model part only has one.

As you have noted, across all optimizers they are always the same, so let's be simple -- users can always extend https://github.com/pytorch/torchtitan/blob/main/torchtitan/protocols/train_spec.py#L55 to do fancier things.

This is also consistent with current checkpoint behavior on LR scheduler:
https://github.com/pytorch/torchtitan/blob/main/torchtitan/components/lr_scheduler.py#L68

A caution is that: since the logging step is happening after `schedulers.step()` call, my understanding is that on step `i`, what's being logged is step `i+1`'s learning rate, instead of step `i`'s.

**Comment by idoh** [comment]
Hey @tianyu-l, I have [updated the PR](https://github.com/pytorch/torchtitan/pull/1413/commits/92f66d4aa62b313524eaa85e19412907d44b8f5f) by simplifying the learning rate logging to only the first one. It's a good point regarding the `i+1` vs `i` step, I did not address this in the commit, let me know if you think it should be addressed.

As for the logging itself, I believe it makes more sense to keep the `log` function minimal, without adding new arguments to the function. When I was thinking about either adding a new argument to `log` I found that there is already [another call](https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/deepseek_v3/train_ds_real.py#L231) in the code that uses this `log` function, and apparently it is already broken because it is missing the recently added `grad_norm` argument. So, I think these additional arguments should just be passed to the `extra_metrics`. The tradeoff here is that then it becomes less straightforward on the [logging for the progress outputs](https://github.com/pytorch/torchtitan/blob/main/torchtitan/components/metrics.py#L408).

Which logging design would you prefer? If we go with adding arguments to `log` then it would require updating it across the codebase in all the places that use `log`. However, if we intend to only have a single place that calls `log` and all the other `train` files should be deprecated then I think we should directly pass the arguments to `log`. Let me know what you prefer and I can update the code accordingly across the repository.

File: torchtitan/components/metrics.py [hunk 1]
```diff
@@ -393,6 +393,10 @@ def log(
             "memory/num_ooms": device_mem_stats.num_ooms,
         }
 
+        if self.lr_schedulers:
+            # Log the learning rate from the first scheduler
```
Comment (tianyu-l) [inline torchtitan/components/metrics.py]: Please add a note that this is logging the LR for step i+1 when logging is done at the end of step i

**Review by tianyu-l — APPROVED** [review]
LGTM, please add a note in the comment.

> When I was thinking about either adding a new argument to log I found that there is already [another call](https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/deepseek_v3/train_ds_real.py#L231) in the code that uses this log function, and apparently it is already broken because it is missing the recently added grad_norm argument. So, I think these additional arguments should just be passed to the extra_metrics. The tradeoff here is that then it becomes less straightforward on the [logging for the progress outputs](https://github.com/pytorch/torchtitan/blob/main/torchtitan/components/metrics.py#L408).

ignore that

File: torchtitan/components/metrics.py (hunk 1 above)
Comment (idoh) [reply torchtitan/components/metrics.py]: Thanks for the feedback. Instead of adding a comment which might confuse some users before looking into the code, I went ahead and fixed it so now the logged learning rate is properly set for step `i` instead of step `i+1`. Furthermore, I removed the unused `optimizers` and `lr_schedulers` in the `MetricsProcessor` as now there is one clear and straightforward way to add new metrics.

File: torchtitan/components/metrics.py [hunk 2]
```diff
@@ -409,6 +407,7 @@ def log(
             f"{color.red}step: {step:2}  "
             f"{color.green}loss: {global_avg_loss:7.4f}  "
             f"{color.orange}grad_norm: {grad_norm:7.4f}  "
+            f"{color.yellow}lr: {lr:.2e}  "
```
Comment (tianyu-l) [inline torchtitan/components/metrics.py]: How important do you think this is? I think we should be very careful in adding fields in terminal prints. I tend to think lr is not very important to put here.

File: torchtitan/components/metrics.py [hunk 3]
```diff
@@ -359,6 +355,7 @@ def log(
         global_avg_loss: float,
         global_max_loss: float,
         grad_norm: float,
+        lr: float,
```
Comment (tianyu-l) [inline torchtitan/components/metrics.py]: maybe just put it in `extra_metrics` instead of adding a new field, if we are not going to print it in terminal.

File: torchtitan/train.py [hunk 4]
```diff
@@ -289,7 +289,6 @@ def __init__(self, job_config: JobConfig):
                 self.model_parts
             )
         )
-        self.metrics_processor.optimizers = self.optimizers
```
Comment (tianyu-l) [inline torchtitan/train.py]: @fduwjj will this break your downstream work?

Comment (fduwjj) [reply torchtitan/train.py]: Right we need this for logging of optimizer's stuff into tb and wandb

Comment (tianyu-l) [reply torchtitan/train.py]: @idoh please revert this change, thanks!

File: torchtitan/components/metrics.py (hunk 1 above)
Comment (idoh) [reply torchtitan/components/metrics.py]: Following our discussion below, I reverted these changes and added the comment about the logging step of `i+1` instead of `i`.

File: torchtitan/components/metrics.py (hunk 2 above)
Comment (idoh) [reply torchtitan/components/metrics.py]: Usually, I like seeing my learning rate, as this can find easy stupid bugs. However, since we also have it in the logger and W&B, it would be even easier to pick that up from the dashboard visually. I reverted this change so it is removed from the terminal change.

File: torchtitan/components/metrics.py (hunk 3 above)
Comment (idoh) [reply torchtitan/components/metrics.py]: Done ✅

File: torchtitan/train.py (hunk 4 above)
Comment (idoh) [reply torchtitan/train.py]: I reverted these change ✅

File: torchtitan/train.py [hunk 5]
```diff
@@ -290,6 +290,7 @@ def __init__(self, job_config: JobConfig):
             )
         )
         self.metrics_processor.optimizers = self.optimizers
```
Comment (tianyu-l) [inline torchtitan/train.py]: Sorry our ask was to revert this line, not the others.
I think we should still log the lr on step i.
You can just update this line to do that
https://github.com/pytorch/torchtitan/blob/main/torchtitan/train.py#L501
without adding `lr_schedulers` to `MetricsProcessor`.

Comment (idoh) [reply torchtitan/train.py]: Done ✅
Let me know if you would like me to remove `lr_schedulers` from from `metrics.py` as it is not being used and probably will not be used in the near future.

**Review by tianyu-l — APPROVED** [review]
LGTM!
